### ISSUE TITLE
excessively long subjectKeyId in org.bouncycastle.cms.KeyTransRecipientId ?

### ISSUE THREAD

**Issue by ralfhauser** [bcgit/bc-java#1380]
The attached mail causes problems on the SMime Decryption.

It seems that the subjectKeyId of the second recipient is excessively long (294 bytes instead of normally approx 20 bytes).

Is there an input check missing here ?

[ticket-2670287-37675.eml.zip](https://github.com/bcgit/bc-java/files/11310011/ticket-2670287-37675.eml.zip)

![grafik](https://user-images.githubusercontent.com/1681677/233986844-65d1ce82-093f-4d7f-bafd-2349a82fa488.png)

(t7028)

**Comment by ralfhauser** [comment]
Some more insights
<img width="462" alt="image" src="https://user-images.githubusercontent.com/1681677/234474504-5e68348b-36db-45b5-9c1a-98c6277e5c8c.png">
<img width="289" alt="image" src="https://user-images.githubusercontent.com/1681677/234474521-b217a4dd-b4b3-4da2-bd44-9e1d2ae5c1a9.png">
<img width="360" alt="image" src="https://user-images.githubusercontent.com/1681677/234474572-8d5136c0-5af7-46f4-908f-6ba6482a0a4c.png">

Parsing error ?

**Comment by peterdettman** [comment]
I have just tested this on latest code and it parses fine, finding a 20 byte subjectKeyId and without errors.

```
    public void testSubKeyIdGithub1380()
    	throws Exception
    {
    	String filename = "Path/To/ticket-2670287-37675.eml";

        Properties props = System.getProperties();

        Session session = Session.getDefaultInstance(props, null);

        MimeMessage msg = new MimeMessage(session, new SharedFileInputStream(filename));

        SMIMEEnvelopedParser m = new SMIMEEnvelopedParser(msg);

        RecipientInformationStore   recipients = m.getRecipientInfos();
    }
```

Can you confirm the specific version of BC you are using and also could you test with SMIMEEnveloped instead of the parser please.

**Comment by ralfhauser** [comment]
Hi Peter,

When I extend your code with the following lines;

                   `Iterator<RecipientInformation>` rIt = recipients.iterator();
                    int ki = 0;
                    while (rIt.hasNext()) {ki++;
						RecipientInformation recipientInformation = (RecipientInformation) rIt
								.next();
						if (recipientInformation instanceof KeyTransRecipientInformation) {
							KeyTransRecipientInformation ktrin = (KeyTransRecipientInformation) recipientInformation;
							RecipientId rid = ktrin.getRID();
							KeyTransRecipientId ktri = (KeyTransRecipientId) rid;
							byte[] ski = ktri.getSubjectKeyIdentifier();
							log.debug(ki+": "+ski.length);
						}
					}`


it turns out that there is more than one subjectKeyId - the output is

1: 20
2: 294

**Comment by peterdettman** [comment]
OK, so the 294 bytes is an encoded SEQUENCE:

```
Sequence
    Sequence
        ObjectIdentifier(1.2.840.113549.1.1.1)
        NULL
    DER Bit String[270, 0] 
```

and that 270 byte BIT STRING is also an encoded SEQUENCE:

```
Sequence
 Integer(19160391302497067434626714016228421851997655162226263015887318605468368628891665938937770336113446563207601554539513769837008303784202016239813021221641142058898772367617743351985020538555361297249922748412184969118105260148481319852670447685101254574556270314721833079665498477532922732652761941888605972463342600810442350987399419570279385283967931308135440687878309318428347753038337520207328081941168699803022177962294162204763397505108168996079460198668510222866515009429159064480929317096544774048992622646806474754128311130152496898281852255006550213801450218917873073167280761134023546341817529331067694724453)
    Integer(65537)
```

So that's presumably an RSA public key and we note that the outer sequence fits an OriginatorPublicKey.

An OriginatorPublicKey should only appear in a KeyAgreeRecipientInfo/OriginatorIdentifierOrKey, and would also need a [CONTEXT 1] tag instead of 0 as it is here. If the encoder was attempting to encode the OriginatorPublicKey instead of wrapping it in a SubjectKeyIdentifier, then they also missed the CONSTRUCTED bit I think.

So the data seems to be a valid ASN.1 encoding, though almost certainly not what the sender intended. It's not obvious that there's a way to restrict the SubjectKeyIdentifier length unless it be a user-specified restriction.

**Comment by ralfhauser** [comment]
Thanks, so as the first ski is presumably referencing the sender's own public key, the second one should reference the recipient's public key.
As you seem to say, that appears to misconstructed, so you would agree that it is not a valid smime encrypted message ?
At least with bouncycastle standard classes, me as the recipient will not be able to decrypt it?
Or is there something missing in the BC smime decryption classes ?

**Comment by peterdettman** [comment]
There are two KeyTransRecipientInfo recipients. These don't reference sender keys, only receiver keys. Any given receiver only needs one of these to be able to decrypt the message and it is normal to send one message to multiple recipients so recipient infos that you can't use should just be gracefully ignored.

Presumably the second KTRI with the long SKI is meaningless to you (or anybody), but if the first one correctly identifies a receiver cert then you should be able to decrypt the session fine.

**Comment by ralfhauser** [comment]
Thanks.
But if the first (20-bytes) SKI doesn't match my key pair as a recipient and the 2nd (meaningless) SKI must be ignored, I cannot decrypt - agreed ?
Or should one try to match the presumable public key with one's own public key ?

**Comment by peterdettman** [comment]
The second one is not technically meaningless, but will not match your certificate. No, you should not try to reinterpret it as an OriginatorPublicKey (that is invalid in a KeyTransRecipientInfo and in any case it would be saying who sent the message not the intended receiver).

Yes, if neither match then you won't be able to decrypt it. If the message was definitely intended for you, then you would have to discuss with the sender as to why they failed to encrypt to you and instead produced this weird construction.

**Comment by ralfhauser** [comment]
If I read "[6.2.1](https://www.rfc-editor.org/rfc/rfc5652#section-6.2.1).  KeyTransRecipientInfo Type" of https://www.rfc-editor.org/rfc/rfc5652#page-22  the attached example does not follow this RFC

**Comment by peterdettman** [comment]
Well the ASN.1 format is consistent with it (which is why it parses successfully), but the actual contents of the SubjectKeyIdentifier looks like it is an encoded OriginatorPublicKey (note the encoded object is just an opaque OCTET STRING as far as the outer ASN.1 parser is concerned), which is at least weird and almost certainly a mistake.

Note this text though:

> When other certificate formats are referenced, the documents that specify the certificate format and their use with the CMS must include details on matching the key identifier to the appropriate certificate field.

...which makes the contents of SubjectKeyIdentifier something of an "open format" so it can only be validated in the context of some certificate format we are trying to match it to (usually X.509 of course), and failure to validate just means no match in that context.

**Comment by ralfhauser** [comment]
[ticketPrivaSphere.cer_till231201.crt.zip](https://github.com/bcgit/bc-java/files/11351434/ticketPrivaSphere.cer_till231201.crt.zip)

Your statement is still confusing to me - if you have the attached certificate - would BC attempt to match it in that specific case (ticket-2670287-37675.eml) or if not what is the justification for not doing it ?

**Comment by peterdettman** [comment]
As far as BC is concerned there are two valid KeyTransRecipientInfo objects in ticket-2670287-37675.eml, both using subjectKeyIdentifier for their RecipientIdentifier. So yes, BC will attempt to match against both of them.

**Comment by ralfhauser** [comment]
> As far as BC is concerned there are two valid KeyTransRecipientInfo objects in ticket-2670287-37675.eml, both using subjectKeyIdentifier for their RecipientIdentifier. So yes, BC will attempt to match against both of them.

But you would agree that it is highly unlikely that the 2nd attempt will ever succeed ?

**Comment by peterdettman** [comment]
Yes I agree. The SubjectKeyIdentifier for an X.509 certificate is a hash of the SubjectPublicKeyInfo, and I doubt anyone is using a hash algorithm with output size 294 bytes.

**Comment by ralfhauser** [comment]
openssl cms -cmsout -noout -print -in ticket-2670287-37675.eml | less

shows details

**Comment by ralfhauser** [comment]
Looking at the details again, it turns out that 

        JcaX509CertSelectorConverter converter = new JcaX509CertSelectorConverter();
        KeyTransRecipientId kRid = null;
        if (rid instanceof KeyTransRecipientId) {
            kRid = (KeyTransRecipientId) rid;
        } else {
            log.warn("rid not a known type " + rid);
        }
        X509CertSelector conv = converter.getCertSelector(kRid);

might be the problem.

before the conversion, the SKI is 20 bytes and
    [-17, -118, 92, 74, -3, 41, 19, -111, -38, 55, 68, -109, 89, 53, -85, -73, 36, -85, 49, -85]

after the conversion with 
   
<img width="550" alt="image" src="https://user-images.githubusercontent.com/1681677/235211083-69b835a0-015c-458f-8cfc-9e4b61ebc6be.png">

it is 22 bytes and

    [4, 20, -17, -118, 92, 74, -3, 41, 19, -111, -38, 55, 68, -109, 89, 53, -85, -73, 36, -85, 49, -85]

So why are the [4,20] prefixed ?

Is
   ConverterTest.testRecipientIdConversion()
with the empty 
   new byte[20]
a realistic test ?

**Comment by ralfhauser** [comment]
Once I did the following patch, my decryption started working 
[1380.patch](https://github.com/bcgit/bc-java/files/11358420/1380.patch)

**Comment by peterdettman** [comment]
Thanks for the patch, it's now applied.

**Comment by dghgit** [comment]
Patch released with 1.74/1.75.
